honeybadgermpc/hbavss.py

```diff
@@ -62,44 +84,78 @@ def multicast(msg):
             for i in range(self.n):
                 send(i, msg)
 
-        commitments, ephemeral_public_key, encrypted_witnesses = loads(avss_msg)
+        commitment, ephemeral_public_key, encrypted_blobs = loads(avss_msg)
         shared_key = pow(ephemeral_public_key, self.private_key)
-        share, witness = SymmetricCrypto.decrypt(
-            str(shared_key).encode("utf-8"), encrypted_witnesses[self.my_id])
-        if self.poly_commit.verify_eval(commitments, self.my_id+1, share, witness):
-            multicast(HbAVSSMessageType.OK)
-        else:
-            logging.error("PolyCommit verification failed.")
-            raise HoneyBadgerMPCError("PolyCommit verification failed.")
-
-        # to handle the case that a (malicious) node sends
-        # an OK message multiple times
+        share_valid = True
+        try:
+            share, witness = SymmetricCrypto.decrypt(
+                str(shared_key).encode(), encrypted_blobs[self.my_id])
+            if self.poly_commit.verify_eval(commitment, self.my_id+1, share, witness):
+                multicast((HbAVSSMessageType.OK, ""))
+            else:
+                multicast((HbAVSSMessageType.IMPLICATE, self.private_key))
+                share_valid = False
+        except Exception:  # TODO specific exceptions
+            multicast((HbAVSSMessageType.IMPLICATE, self.private_key))
+            share_valid = False
+
+        # RECEIVE LOOP
         ok_set = set()
-        while len(ok_set) < 2 * self.t + 1:
-            sender, avss_msg = await recv()  # First value is the `sid`
-            if avss_msg == HbAVSSMessageType.OK and sender not in ok_set:
+        recovery_shares = []
+        sent_recovery = False
+        output = False
+        while True:
+            if len(ok_set) == 2 * self.t + 1 and share_valid and not output:
+                share_int = int(share)
+                self.output_queue.put_nowait((dealer_id, avss_id, share_int))
+                output = True
+            if len(recovery_shares) == self.t + 1 and not output:
```

Comment:
Is there ever a case where both of these `if` statements are supposed to be true? If not, we should make this second `if` and `elif` to make this clearer.